azext_edge/edge/providers/adr/validator.py

```diff
@@ -0,0 +1,636 @@
+
+        # Handle mcr.microsoft.com specific logic if needed, or generic OCI
+        # MCR redirects to data endpoints, so standard requests usually work if we follow redirects.
+
+        base_url = f"https://{registry}/v2/{repository}"
+        headers = {
+            "Accept": "application/vnd.oci.image.manifest.v1+json, application/vnd.docker.distribution.manifest.v2+json"
+        }
+
+        # 2. Get Manifest
+        manifest_url = f"{base_url}/manifests/{tag}"
+
+        # Note: For private registries, we'd need authentication (Bearer token).
+        # MCR public images usually allow anonymous pull but might require a token handshake.
+        # For simplicity in this draft, we'll assume public access or implement a basic token flow.
+
+        token = ConnectorMetadataValidator._get_auth_token(registry, repository)
+        if token:
+            headers["Authorization"] = f"Bearer {token}"
+
+        response = requests.get(manifest_url, headers=headers, timeout=30)
+        if response.status_code != 200:
+            raise ValidationError(f"Failed to fetch manifest for {image_ref}: {response.status_code} {response.text}")
+
+        manifest = response.json()
+
+        # 3. Find the connector-metadata.json layer
+        # The connector metadata is stored as a file named "connector-metadata.json" in the OCI artifact
+        target_digest = None
+
+        layers = manifest.get("layers", [])
+
+        # Strategy 1: Look for layer with title annotation containing "connector-metadata.json"
+        for layer in layers:
+            annotations = layer.get("annotations", {})
+            title = annotations.get("org.opencontainers.image.title", "")
+
+            # Match if title ends with connector-metadata.json
+            # (handles paths like "azure_iot_operations_rest_connector/connector-metadata.json")
+            if title.endswith("connector-metadata.json"):
+                target_digest = layer["digest"]
+                logger.info(f"Found connector-metadata.json layer: {target_digest}")
+                break
+
+        # Strategy 2: If not found by title, look for application/json media type
+        if not target_digest:
+            for layer in layers:
+                media_type = layer.get("mediaType", "")
+                if "json" in media_type.lower():
+                    target_digest = layer["digest"]
+                    logger.info(f"Found JSON layer by media type: {target_digest}")
+                    break
+
+        # Strategy 3: Try the first layer as fallback
+        if not target_digest and len(layers) > 0:
+            target_digest = layers[0]["digest"]
+            logger.warning(f"Using first layer as fallback: {target_digest}")
+
+        if not target_digest:
+            raise ValidationError(
+                f"Could not find connector-metadata.json layer in {image_ref}. "
+                f"Manifest has {len(layers)} layers but none match expected structure."
+            )
+
```

Comment:
We can't rely on file name and there's a specific annotation we need to check for connectors. Ping me separately for the details.